File: jogo.py
def cria_placar():
    placar = [0, 0]
    return placar

def cria_tabuleiro():
    tabuleiro = [[0 for _ in range(7)] for _ in range(6)]
    return tabuleiro

def print_tabuleiro(tabuleiro):
    for linha in reversed(tabuleiro):
        print(linha)

def verifica_vitoria(tabuleiro, peca):
    for linha in range(6):
        for col in range(4):
            if tabuleiro[linha][col] == peca and tabuleiro[linha][col+1] == peca and tabuleiro[linha][col+2] == peca and tabuleiro[linha][col+3] == peca:
                return True

    for col in range(7):
        for linha in range(3):
            if tabuleiro[linha][col] == peca and tabuleiro[linha+1][col] == peca and tabuleiro[linha+2][col] == peca and tabuleiro[linha+3][col] == peca:
                return True

    for linha in range(3):
        for col in range(4):
            if tabuleiro[linha][col] == peca and tabuleiro[linha+1][col+1] == peca and tabuleiro[linha+2][col+2] == peca and tabuleiro[linha+3][col+3] == peca:
                return True

    for linha in range(3,6):
        for col in range(4):
            if tabuleiro[linha][col] == peca and tabuleiro[linha-1][col+1] == peca and tabuleiro[linha-2][col+2] == peca and tabuleiro[linha-3][col+3] == peca:
                return True

    return False

def procura_linha(tabuleiro, coluna):
    for linha in range(6):
        if tabuleiro[linha][coluna] == 0:
            return linha
    return None

def jogada(tabuleiro, jogador):
    while True:
        entrada = input()
        
        if entrada == 'd':
            print(f'O jogador {jogador} desistiu.')
            return "desistiu"

        if entrada.isdigit():
            col = int(entrada) - 1
            if 0 <= col <= 6:
                if tabuleiro[5][col] == 0:
                    linha = procura_linha(tabuleiro, col)
                    tabuleiro[linha][col] = jogador
                    return jogador
                else:
                    print("Coluna cheia. Tente outra.")
            else:
                print("Escolha uma coluna entre 1 e 7.")
        else:
            print("Entrada inválida. Tente novamente...")

def jogo():
    tabuleiro = cria_tabuleiro()
    turno = 0
    acabou = False

    while not acabou:
        print_tabuleiro(tabuleiro)
        jogador = 1 if turno % 2 == 0 else 2
        resultado = jogada(tabuleiro, jogador)

        if resultado == "desistiu":
            return 0
        elif verifica_vitoria(tabuleiro, jogador):
            print_tabuleiro(tabuleiro)
            print(f'O jogador {jogador} venceu!')
            return jogador
        else:
            turno += 1

def start_game():
    placar = cria_placar()
    continuar = True

    while continuar:
        vencedor = jogo()
        if vencedor == 0: break
        placar[vencedor - 1] += 1
        print(f'Placar - 1º jogador {placar[0]} x {placar[1]} 2º jogador')

        resposta = input("Deseja jogar novamente? (s/n): ").lower()
        if resposta != 's':
            continuar = False

    print("Fim")

start_game()

File: jogo.clj
(defn cria-tabuleiro []
  (vec (repeat 6 (vec (repeat 7 0)))))

(defn imprime-tabuleiro [tabuleiro]
  (doseq [linha (reverse tabuleiro)]
    (println linha)))

(defn procura-linha [tabuleiro col]
  (loop [linha 0]
    (if (< linha 6)
      (if (= 0 (get-in tabuleiro [linha col]))
        linha
        (recur (inc linha)))
      nil)))

(defn jogada [tabuleiro jogador]
  (loop [tabuleiro-atual tabuleiro]
    (imprime-tabuleiro tabuleiro)
    (let [entrada (read-line)
          resultado (cond
                      (= entrada "d")
                      (do
                        (println (str "O jogador " jogador " desistiu!"))
                        (System/exit 0))
                      
                      (re-matches #"\d" entrada)
                      (let [col (dec (Integer/parseInt entrada))]
                        (if (and (>= col 0) (< col 7))
                          (let [linha (procura-linha tabuleiro-atual col)]
                            (if linha
                              (assoc-in tabuleiro-atual [linha col] jogador)
                              (do
                                (println "Coluna cheia, tente outra.")
                                tabuleiro-atual)))
                          (do
                            (println "Entrada inválida, escolha de 1 a 7.")
                            tabuleiro-atual)))
                      
                      :else (do
                              (println "Entrada inválida, tente novamente.")
                              tabuleiro-atual))]

      (if (= resultado tabuleiro-atual)
        (recur tabuleiro-atual)
        resultado))))

(defn verifica-vitoria [tabuleiro peca]
  (or
    (some true?
      (map (fn [linha]
             (some (fn [col]
                     (and (= (get-in tabuleiro [linha col]) peca)
                          (= (get-in tabuleiro [linha (+ col 1)]) peca)
                          (= (get-in tabuleiro [linha (+ col 2)]) peca)
                          (= (get-in tabuleiro [linha (+ col 3)]) peca)))
                   (range 4)))
           (range 6)))
    
    (some true?
      (map (fn [col]
             (some (fn [linha]
                     (and (= (get-in tabuleiro [linha col]) peca)
                          (= (get-in tabuleiro [(+ linha 1) col]) peca)
                          (= (get-in tabuleiro [(+ linha 2) col]) peca)
                          (= (get-in tabuleiro [(+ linha 3) col]) peca)))
                   (range 3)))
           (range 7)))
    
    (some true?
      (map (fn [linha]
             (some (fn [col]
                     (and (= (get-in tabuleiro [linha col]) peca)
                          (= (get-in tabuleiro [(+ linha 1) (+ col 1)]) peca)
                          (= (get-in tabuleiro [(+ linha 2) (+ col 2)]) peca)
                          (= (get-in tabuleiro [(+ linha 3) (+ col 3)]) peca)))
                   (range 4)))
           (range 3)))
    
    (some true?
      (map (fn [linha]
             (some (fn [col]
                     (and (= (get-in tabuleiro [linha col]) peca)
                          (= (get-in tabuleiro [(- linha 1) (+ col 1)]) peca)
                          (= (get-in tabuleiro [(- linha 2) (+ col 2)]) peca)
                          (= (get-in tabuleiro [(- linha 3) (+ col 3)]) peca)))
                   (range 4)))
           (range 3 6)))))

(defn jogar []
  (let [tabuleiro (cria-tabuleiro)]
    (loop [turno 0
           tabuleiro-atual tabuleiro]
      (let [jogador (if (zero? (mod turno 2)) 1 2)
            tabuleiro-atualizado (jogada tabuleiro-atual jogador)]
        (cond
          (verifica-vitoria tabuleiro-atualizado jogador)
          (do
            (imprime-tabuleiro tabuleiro-atualizado)
            (println (str "O jogador " jogador " venceu!"))
            jogador)

          :else
          (recur (inc turno) tabuleiro-atualizado))))))

(defn start-game
  ([]
   (start-game [0 0]))
  ([placar]
   (let [continuar true]
   (while continuar
     (let [vencedor (jogar)]
       (if (and vencedor (<= vencedor 2))
         (do
           (let [novo-placar (update-in placar [(dec vencedor)] inc)]
             (println (str "Placar: 1º Jogador " (first novo-placar) " - " (second novo-placar) " 2º Jogador"))

             (println "Você deseja continuar jogando? (s/n):")
             (let [resposta (clojure.string/lower-case (read-line))]
               (if (= resposta "s")
                 (start-game novo-placar)
                 (do
                   (println "Fim de jogo!")
                   (System/exit 0))))))))))))

(start-game)

